### ISSUE TITLE
Upgrading Waypoint proxies to 1.24.0 causes brief outage with filter_chain_not_found errors

### ISSUE THREAD

**Issue by adam-harries-dct** [istio/istio#53883]
### Is this the right place to submit this?

- [X] This is not a security vulnerability or a crashing bug
- [X] This is not a question about how to use Istio

### Bug Description

I am running Istio in Ambient mode and have followed the upgrade guide to upgrade Istio from version 1.23.2 to version 1.24.0. The steps I followed in order were:

1. Upgrade istio-base helm chart
2. Install Istiod-1.24.0 Helm chart, alongside 1.23.2, setting `istio.io/rev: 1-24-0`
3. Upgrade istio-cni helm chart
4. Restart all meshed workloads (not necessary but did so for good measure)
5. Upgrade istio-ztunnel helm chart, setting `istio.io/rev: 1-24-0`
6. Edit Waypoint Gateway proxies to change label `istio.io/rev: 1-23-2` to `istio.io/rev: 1-24-0`

Everything continued to work as normal until step 5. As soon as I upgraded the Waypoint proxies, all ingress traffic to the cluster started failing with the same type of error from the Waypoint proxy:

```
{
  bytes_received 0
  bytes_sent 0
  downstream_local_address 10.0.181.232:8877
  downstream_remote_address 10.0.1.177:58622
  http {
    status_code 0
  }
  response_code 0
  response_code_details filter_chain_not_found
  response_flags NR
  start_time 2024-11-13T16:51:29.341Z
}
```

This lasted for approximately 10 minutes until everything started working as normal again. Clearly _something_ happened in the background that fixed the cause of the issue but I'm struggling to work out what it was and how to avoid this happening on a production cluster.

Please note that the helm chart upgrades were deployed via ArgoCD rather than via the Helm CLI.

### Version

```Text
➜ istioctl version
client version: 1.24.0
istiod version: 1.23.2
istiod version: 1.23.2
istiod version: 1.23.2
istiod version: 1.24.0
istiod version: 1.24.0
istiod version: 1.24.0
data plane version: 1.24.0 (10 proxies)

➜ kubectl version        
Client Version: v1.30.2
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.29.9

➜ istioctl proxy-status
NAME                               CLUSTER        CDS              LDS              EDS              RDS         ECDS        ISTIOD                             VERSION
waypoint-7445477dcb-672l2.load     Kubernetes     SYNCED (36s)     SYNCED (36s)     SYNCED (20m)     IGNORED     IGNORED     istiod-1-24-0-5b48fc56d7-v8pds     1.24.0
waypoint-7445477dcb-cp4tl.load     Kubernetes     SYNCED (36s)     SYNCED (36s)     SYNCED (20m)     IGNORED     IGNORED     istiod-1-24-0-5b48fc56d7-69dq6     1.24.0
waypoint-7445477dcb-v6rcv.load     Kubernetes     SYNCED (36s)     SYNCED (36s)     SYNCED (20m)     IGNORED     IGNORED     istiod-1-24-0-5b48fc56d7-v8pds     1.24.0
ztunnel-49ff7.istio-system         Kubernetes     IGNORED          IGNORED          IGNORED          IGNORED     IGNORED     istiod-1-24-0-5b48fc56d7-v8pds     1.24.0
ztunnel-8swgn.istio-system         Kubernetes     IGNORED          IGNORED          IGNORED          IGNORED     IGNORED     istiod-1-24-0-5b48fc56d7-v8pds     1.24.0
ztunnel-fwkpw.istio-system         Kubernetes     IGNORED          IGNORED          IGNORED          IGNORED     IGNORED     istiod-1-24-0-5b48fc56d7-69dq6     1.24.0
ztunnel-n2khr.istio-system         Kubernetes     IGNORED          IGNORED          IGNORED          IGNORED     IGNORED     istiod-1-24-0-5b48fc56d7-v8pds     1.24.0
ztunnel-s9k9d.istio-system         Kubernetes     IGNORED          IGNORED          IGNORED          IGNORED     IGNORED     istiod-1-24-0-5b48fc56d7-j7fx4     1.24.0
ztunnel-vfs55.istio-system         Kubernetes     IGNORED          IGNORED          IGNORED          IGNORED     IGNORED     istiod-1-24-0-5b48fc56d7-v8pds     1.24.0
ztunnel-zfxm9.istio-system         Kubernetes     IGNORED          IGNORED          IGNORED          IGNORED     IGNORED     istiod-1-24-0-5b48fc56d7-j7fx4     1.24.0
```

### Additional Information

istio-namespace: istio-system
full-secrets: false
timeout (mins): 30
include: {  }
exclude: { Namespaces: kube-node-lease,kube-public,kube-system,local-path-storage }
end-time: 2024-11-13 17:16:52.371509 +0000 UTC

Cluster endpoint: https://##################
CLI version:
version.BuildInfo{Version:"1.24.0", GitRevision:"Homebrew", GolangVersion:"go1.23.3", BuildStatus:"Homebrew", GitTag:"1.24.0"}

The following Istio control plane revisions/versions were found in the cluster:
Revision default:
no running Istio pods in namespace "istio-system"

Revision 1-23-2:
&version.MeshInfo{version.ServerInfo{Component:"istiod", Revision:"1-23-2", Info:version.BuildInfo{Version:"1.23.2", GitRevision:"da0b3f29e9f9eef387691bdf5870a6d665c704c9", GolangVersion:"", BuildStatus:"Clean", GitTag:"1.23.2"}}, version.ServerInfo{Component:"istiod", Revision:"1-23-2", Info:version.BuildInfo{Version:"1.23.2", GitRevision:"da0b3f29e9f9eef387691bdf5870a6d665c704c9", GolangVersion:"", BuildStatus:"Clean", GitTag:"1.23.2"}}, version.ServerInfo{Component:"istiod", Revision:"1-23-2", Info:version.BuildInfo{Version:"1.23.2", GitRevision:"da0b3f29e9f9eef387691bdf5870a6d665c704c9", GolangVersion:"", BuildStatus:"Clean", GitTag:"1.23.2"}}}

Revision 1-24-0:
&version.MeshInfo{version.ServerInfo{Component:"istiod", Revision:"1-24-0", Info:version.BuildInfo{Version:"1.24.0", GitRevision:"8825a6b7f8c9a2d66005a5f8b64e98aaee0dda99", GolangVersion:"", BuildStatus:"Clean", GitTag:"1.24.0"}}, version.ServerInfo{Component:"istiod", Revision:"1-24-0", Info:version.BuildInfo{Version:"1.24.0", GitRevision:"8825a6b7f8c9a2d66005a5f8b64e98aaee0dda99", GolangVersion:"", BuildStatus:"Clean", GitTag:"1.24.0"}}, version.ServerInfo{Component:"istiod", Revision:"1-24-0", Info:version.BuildInfo{Version:"1.24.0", GitRevision:"8825a6b7f8c9a2d66005a5f8b64e98aaee0dda99", GolangVersion:"", BuildStatus:"Clean", GitTag:"1.24.0"}}}

The following proxy revisions/versions were found in the cluster:
Revision 1-24-0: Versions {1.24.0}

Fetching logs for the following containers:
###################### (obfuscated for security)

Fetching Istio control plane information from cluster.

Fetching CNI logs from cluster.

Running Istio analyze on all namespaces and report as below:
Analysis Report:
Info [IST0102] (Namespace datadog) The namespace is not enabled for Istio injection. Run 'kubectl label namespace datadog istio-injection=enabled' to enable it, or 'kubectl label namespace datadog istio-injection=disabled' to explicitly mark it as not needing injection.
Info [IST0102] (Namespace default) The namespace is not enabled for Istio injection. Run 'kubectl label namespace default istio-injection=enabled' to enable it, or 'kubectl label namespace default istio-injection=disabled' to explicitly mark it as not needing injection.
Info [IST0102] (Namespace emissary-system) The namespace is not enabled for Istio injection. Run 'kubectl label namespace emissary-system istio-injection=enabled' to enable it, or 'kubectl label namespace emissary-system istio-injection=disabled' to explicitly mark it as not needing injection.
Info [IST0102] (Namespace keda) The namespace is not enabled for Istio injection. Run 'kubectl label namespace keda istio-injection=enabled' to enable it, or 'kubectl label namespace keda istio-injection=disabled' to explicitly mark it as not needing injection.
Info [IST0102] (Namespace kube-node-lease) The namespace is not enabled for Istio injection. Run 'kubectl label namespace kube-node-lease istio-injection=enabled' to enable it, or 'kubectl label namespace kube-node-lease istio-injection=disabled' to explicitly mark it as not needing injection.
Info [IST0102] (Namespace kube-public) The namespace is not enabled for Istio injection. Run 'kubectl label namespace kube-public istio-injection=enabled' to enable it, or 'kubectl label namespace kube-public istio-injection=disabled' to explicitly mark it as not needing injection.
Info [IST0102] (Namespace kube-system) The namespace is not enabled for Istio injection. Run 'kubectl label namespace kube-system istio-injection=enabled' to enable it, or 'kubectl label namespace kube-system istio-injection=disabled' to explicitly mark it as not needing injection.
Info [IST0118] (Service load/admin-portal) Port name  (port: 80, targetPort: 8080) doesn't follow the naming convention of Istio port.
Creating an archive at /Users/###############
Time used for creating the tar file is 387.589959ms.
Cleaning up temporary files in /var/folders/km/txfjl8cj3j15s5wzrnv6l36r0000gq/T/bug-report.
Done.

**Comment by keithmattix** [comment]
At what point did you install the new istiod helm chart?

**Comment by adam-harries-dct** [comment]
> At what point did you install the new istiod helm chart?

Good point, I forgot to include that step. I installed Istiod 1.24.0 after istio-base was ugpraded. It was installed alongside istiod-1.23.2.

**Comment by adam-harries-dct** [comment]
I've conducted some further tests and _think_ I have found a stable upgrade path; rather than changing the `revision` label on Waypoint proxy (named `waypoint`) from `istio.io/rev=1-23-2` to `istio.io/rev=1-24-0`, I instead deployed a new  Waypoint proxy, named `waypoint-1-24-0`, alongside the existing one with the label `istio.io/rev=1-24-0`. I allowed the new Waypoint pods to spin up then edited the label on the namespace (I use a single Waypoint per namespace) from `istio.io/use-waypoint=waypoint` to `istio.io/use-waypoint=waypoint-1-24-0`. This approached seemed stable and didn't appear to cause any downtime.

Further, I found that, for my initial approach where I edited the `revision` of the existing Waypoint, the issue seemed to resolve when I scaled down the `istio-1-23-2` deployment to 0 pods. I.e. it was the presence of both `1.23.2` and `1.24.0` istiod pods that seemed to cause the issue.
